File: src/lib/paddock.ts
const API_BASE_URL = (process.env.NEXT_PUBLIC_API_URL || '/api').replace(/\/+$/, '');

export type cropType = 'default' | 'Grains' | 'Legumes' | 'Fruit' |'Oil Seeds' |'Root Crops' | 'Tropical'| 'Other';

export interface CreatePaddockRequest {
  paddock_name: string | null;
  crop_type: cropType;
  area: number | null;
  plant_date: string;  // Required
}

export interface CreatePaddockResponse {
  success: boolean;
  message: string;
}

export interface PaddockApiError {
  success: false;
  message: string;
}

export interface UpdatePaddockRequest {
  paddock_name: string;
  crop_type: cropType;
  area: number;
}

export interface UpdatePaddockResponse {
  success: boolean;
  message: string;
  paddock?: {
    paddock_id: number;
    paddock_name: string;
    crop_type: cropType;
    area: number
  };
}

export interface GetDevicesPaddockResponse {
    success: boolean;
    message: string;
    devices: Array<{
      node_id: string;
      node_name: string;
      battery: number;
    }>;
}

export interface SensorAveragesResponse {
  success: boolean;
  message: string;
  paddock_id?: number;
  paddock_name?: string;
  nodes_count?: number;
  nodes_with_readings?: number;
  sensor_averages?: {
    [key: string]: number;
  };
  sensor_details?: {
    [key: string]: {
      average: number;
      min: number;
      max: number;
      count: number;
    };
  };
}

export async function getPaddockDevices(
    paddockId: string,
    token: string
  ): Promise<GetDevicesPaddockResponse | PaddockApiError> {
    try {
      const response = await fetch(`${API_BASE_URL}/paddock/${paddockId}/devices`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
      });
  
      const data = await response.json();
  
      if (!response.ok) {
        return {
          success: false,
          message: data.message || 'Failed to fetch devices',
          devices: [],
        };
      }
        return {
        success: true,
        message: data.message || 'Devices fetched successfully',
        devices: data.devices || [],
      };

    } catch (error) {
      console.error('Error fetching paddock devices:', error);
      return {
        success: false,
        message: 'An error occurred while fetching devices',
        devices: [],
      };
    }
}

/**
 * Gets sensor averages for a paddock
 * @param paddockId - ID of the paddock
 * @param token - JWT authentication token
 * @returns Promise with sensor averages data
 */
export async function getPaddockSensorAverages(
  paddockId: string,
  token: string
): Promise<SensorAveragesResponse | PaddockApiError> {
  try {
    const response = await fetch(`${API_BASE_URL}/paddock/${paddockId}/sensor-averages`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
    });

    const data = await response.json();

    if (!response.ok) {
      return {
        success: false,
        message: data.message || 'Failed to fetch sensor averages',
      };
    }

    return {
      success: true,
      message: data.message || 'Sensor averages fetched successfully',
      paddock_id: data.paddock_id,
      paddock_name: data.paddock_name,
      nodes_count: data.nodes_count,
      nodes_with_readings: data.nodes_with_readings,
      sensor_averages: data.sensor_averages || {},
      sensor_details: data.sensor_details || {},
    };
  } catch (error) {
    console.error('Error fetching sensor averages:', error);
    return {
      success: false,
      message: 'An error occurred while fetching sensor averages',
    };
  }
}

/**
 * Creates a new paddock
 * @param paddockName - Name of the paddock (optional)
 * @param cropType - Type of the paddock
 * @param area - Area in hectares (optional, as string from input)
 * @param plant_date - Plant date (REQUIRED, as string from datetime-local)
 * @param token - JWT authentication token
 * @returns Promise with the API response
 */
export async function createPaddock(
  paddockName: string | null,
  cropType: cropType,
  area: string,
  plant_date: string,
  soilType: string,
  token: string
): Promise<CreatePaddockResponse> {
  try {
    // Validate plant_date is not empty
    if (!plant_date || plant_date.trim() === '') {
      return {
        success: false,
        message: 'Plant date is required',
      };
    }

    // Convert area to number or null
    const areaValue = area && area.trim() !== '' ? parseInt(area, 10) : null;

    const response = await fetch(`${API_BASE_URL}/paddock/create`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify({
        paddock_name: paddockName || null,
        crop_type: cropType,
        area: areaValue,
        date_plant: plant_date,  // Send as-is, required
        soil_type: soilType
      }),
    });

    const data = await response.json();

    if (!response.ok) {
      return {
        success: false,
        message: data.message || data.detail || 'Failed to create paddock',
      };
    }

    return {
      success: true,
      message: data.message || 'Paddock created successfully'
    };
  } catch (error) {
    console.error('Error creating paddock:', error);
    return {
      success: false,
      message: 'An error occurred. Please try again.',
    };
  }
}

/**
 * Gets all paddocks for the current user
 * @param token - JWT authentication token
 * @returns Promise with array of paddocks
 */
export async function getPaddocks(token: string) {
  try {
    const response = await fetch(`${API_BASE_URL}/paddock/list`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
    });

    const data = await response.json();

    if (!response.ok) {
      return {
        success: false,
        message: data.message || 'Failed to fetch paddocks',
        paddocks: [],
      };
    }

    return {
      success: true,
      paddocks: data.paddocks || [],
    };
  } catch (error) {
    console.error('Error fetching paddocks:', error);
    return {
      success: false,
      message: 'An error occurred while fetching paddocks',
      paddocks: [],
    };
  }
}

/**
 * Updates a paddock's name
 * @param paddockId - ID of the paddock to update
 * @param paddockName - New name for the paddock
 * @param cropType - Type of the paddock
 * @param area - Area in hectares
 * @param token - JWT authentication token
 * @returns Promise with the API response
 */
export async function updatePaddockName(
  paddockId: string,
  paddockName: string,
  cropType: cropType,
  area: number,
  token: string
): Promise<UpdatePaddockResponse> {
  try {
    const response = await fetch(`${API_BASE_URL}/paddock/${paddockId}`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify({
        paddock_name: paddockName,
        crop_type: cropType,
        area: area
      }),
    });

    const data = await response.json();

    if (!response.ok) {
      return {
        success: false,
        message: data.message || data.detail || 'Failed to update paddock',
      };
    }

    return {
      success: true,
      message: data.message || 'Paddock updated successfully',
      paddock: data.paddock,
    };
  } catch (error) {
    console.error('Error updating paddock:', error);
    return {
      success: false,
      message: 'An error occurred. Please try again.',
    };
  }
}

/**
 * Deletes a paddock
 * @param paddockId - ID of the paddock to delete
 * @param token - JWT authentication token
 * @returns Promise with the API response
 */
export async function deletePaddock(paddockId: string | number, token: string) {
  try {
    const response = await fetch(`${API_BASE_URL}/paddock/${paddockId}`, {
      method: 'DELETE',
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });

    const data = await response.json();
    return data;
  } catch (error) {
    console.error('Error deleting paddock:', error);
    return {
      success: false,
      message: 'Failed to delete paddock',
    };
  }
}

File: netlink/routes/paddock.py
from fastapi import Depends, status, APIRouter
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session
from sqlalchemy import func
from typing import Optional, Annotated
from pydantic import BaseModel
import db.db_models as db_models
from db.init import engine, get_db
from middleware.middleware import get_current_user
from typing import List
from datetime import datetime

db_dependency = Annotated[Session, Depends(get_db)]
db_models.Base.metadata.create_all(bind=engine)

router = APIRouter(prefix="/paddock")

class PaddockCreate(BaseModel):
    paddock_name: Optional[str] = None

class PaddockUpdate(BaseModel):
    paddock_name: str

def generate_paddock_name(count: int) -> str:
    """Generate a unique paddock name based on count"""
    counter = count + 1
    base_name = f"Paddock-{counter}"
    return base_name

@router.post("/create", tags=["paddock"], description="Creates a paddock.")
async def create_paddock(
    paddock_to_create: PaddockCreate,
    db: Session = Depends(get_db),
    current_user: db_models.user = Depends(get_current_user)
):
    # Generate name if not provided, otherwise use provided name
    if not paddock_to_create.paddock_name:
        paddock_count = db.query(db_models.paddock).filter(
            db_models.paddock.user_id == current_user.user_id
        ).count()
        paddock_name = generate_paddock_name(paddock_count)
    else:
        paddock_name = paddock_to_create.paddock_name
    
    # Check if a paddock with the same name already exists for this user
    existing_paddock = db.query(db_models.paddock).filter(
        db_models.paddock.user_id == current_user.user_id,
        db_models.paddock.paddock_name == paddock_name
    ).first()
    
    if existing_paddock:
        return JSONResponse(
            content={
                "success": False,
                "message": f"A paddock named '{paddock_name}' already exists"
            },
            status_code=status.HTTP_409_CONFLICT
        )
    
    new_paddock = db_models.paddock(
        paddock_name=paddock_name,
        user_id=current_user.user_id
    )
    
    db.add(new_paddock)
    db.commit()
    db.refresh(new_paddock)
    
    return JSONResponse(
        content={
            "success": True,
            "message": "Paddock created successfully"
        },
        status_code=status.HTTP_201_CREATED
    )

@router.get("/list", tags=["paddock"], description="Gets all paddocks for the current user.")
async def list_paddocks(
    db: Session = Depends(get_db),
    current_user: db_models.user = Depends(get_current_user)
):
    """
    Retrieves all paddocks belonging to the authenticated user
    """
    try:
        # Query all paddocks for the current user
        paddocks = db.query(db_models.paddock).filter(
            db_models.paddock.user_id == current_user.user_id
        ).order_by(db_models.paddock.paddock_id).all()
        
        # Convert to list of dictionaries
        paddocks_list = [
            {
                "paddock_id": paddock.paddock_id,
                "paddock_name": paddock.paddock_name
            }
            for paddock in paddocks
        ]
        
        return JSONResponse(
            content={
                "success": True,
                "paddocks": paddocks_list,
                "count": len(paddocks_list)
            },
            status_code=status.HTTP_200_OK
        )
    
    except Exception as e:
        return JSONResponse(
            content={
                "success": False,
                "message": "Failed to retrieve paddocks",
                "error": str(e)
            },
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
        )
    
@router.get("/{paddock_id}", tags=["paddock"], description="Get details for a specific paddock.")
async def get_paddock(
    paddock_id: int,
    db: Session = Depends(get_db),
    current_user: db_models.user = Depends(get_current_user)
):
    """
    Retrieves detailed information about a paddock for the authenticated user.
    """
    try:
        paddock = db.query(db_models.paddock).filter(
            db_models.paddock.user_id == current_user.user_id,
            db_models.paddock.paddock_id == paddock_id
        ).first()

        if not paddock:
            return JSONResponse(
                content={
                    "success": False,
                    "message": "Paddock not found"
                },
                status_code=status.HTTP_404_NOT_FOUND
            )

        # Return paddock details
        paddock_data = {
            "paddock_id": paddock.paddock_id,
            "paddock_name": paddock.paddock_name,
            "user_id": paddock.user_id,
            "created_at": paddock.created_at.isoformat() if paddock.created_at else None, #created at field gives error
            # Add other fields if you have more info
        }

        return JSONResponse(
            content={
                "success": True,
                "paddock": paddock_data
            },
            status_code=status.HTTP_200_OK
        )

    except Exception as e:
        return JSONResponse(
            content={
                "success": False,
                "message": "Failed to retrieve paddock",
                "error": str(e)
            },
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
        )


@router.get("/{paddock_id}/devices", tags=["paddock"], description="Gets all devices for a paddock")
async def get_paddock_devices(
    paddock_id: int,
    db: Session = Depends(get_db),
    current_user: db_models.user = Depends(get_current_user)
):
    """
    Retrieves all devices belonging to a specific paddock
    for the authenticated user.
    """
    try:
        # Ensure paddock belongs to current user
        paddock = db.query(db_models.paddock).filter(
            db_models.paddock.user_id == current_user.user_id,
            db_models.paddock.paddock_id == paddock_id
        ).first()

        if not paddock:
            return JSONResponse(
                content={
                    "success": False,
                    "message": "Paddock not found"
                },
                status_code=status.HTTP_404_NOT_FOUND
            )

        # Get all devices linked to this paddock
        device_list = db.query(db_models.active_node).filter(
            db_models.active_node.paddock_id == paddock.paddock_id
        ).all()

        # If no devices found, return an empty list (not an error)
        if not device_list:
            return JSONResponse(
                content={
                    "success": True,
                    "devices": [],
                    "count": 0
                },
                status_code=status.HTTP_200_OK
            )

        # Convert SQLAlchemy objects to plain dicts
        devices_data = [
            {
                "node_id": device.node_id,
                "node_name": device.node_name
            }
            for device in device_list
        ]

        return JSONResponse(
            content={
                "success": True,
                "devices": devices_data
            },
            status_code=status.HTTP_200_OK
        )

    except Exception as e:
        return JSONResponse(
            content={
                "success": False,
                "message": "Failed to retrieve paddock devices",
                "error": str(e)
            },
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
        )

@router.get("/{paddock_id}/sensor-averages", tags=["paddock"], description="Get average sensor readings from all nodes in a paddock")
async def get_paddock_sensor_averages(
    paddock_id: int,
    db: Session = Depends(get_db),
    current_user: db_models.user = Depends(get_current_user)
):
    """
    Retrieves the most recent sensor readings from each node in a paddock
    and returns the average values for each sensor type (pH, temp, etc.).
    """
    try:
        # Verify paddock belongs to current user
        paddock = db.query(db_models.paddock).filter(
            db_models.paddock.user_id == current_user.user_id,
            db_models.paddock.paddock_id == paddock_id
        ).first()

        if not paddock:
            return JSONResponse(
                content={
                    "success": False,
                    "message": "Paddock not found"
                },
                status_code=status.HTTP_404_NOT_FOUND
            )

        # Get all nodes in this paddock
        nodes = db.query(db_models.active_node).filter(
            db_models.active_node.paddock_id == paddock_id
        ).all()

        if not nodes:
            return JSONResponse(
                content={
                    "success": True,
                    "message": "No nodes found in this paddock",
                    "paddock_id": paddock_id,
                    "paddock_name": paddock.paddock_name,
                    "nodes_count": 0,
                    "sensor_averages": {},
                },
                status_code=status.HTTP_200_OK
            )

        node_ids = [node.node_id for node in nodes]

        # Subquery to get the most recent capture_id for each node and reading_type combination
        latest_captures_query = db.query(
            db_models.reading.node_id,
            db_models.reading.reading_type,
            func.max(db_models.reading.capture_id).label('latest_capture_id')
        ).filter(
            db_models.reading.node_id.in_(node_ids)
        ).group_by(
            db_models.reading.node_id,
            db_models.reading.reading_type
        )
        
        latest_captures_results = latest_captures_query.all()
        
        latest_captures = latest_captures_query.subquery()

        # Get readings for the most recent capture of each sensor type per node
        recent_readings = db.query(
            db_models.reading.reading_type,
            db_models.reading.reading_val,
            db_models.reading.node_id,
            db_models.reading.capture_id
        ).join(
            latest_captures,
            (db_models.reading.node_id == latest_captures.c.node_id) &
            (db_models.reading.reading_type == latest_captures.c.reading_type) &
            (db_models.reading.capture_id == latest_captures.c.latest_capture_id)
        ).all()

        if not recent_readings:
            return JSONResponse(
                content={
                    "success": True,
                    "message": "No sensor readings found for nodes in this paddock",
                    "paddock_id": paddock_id,
                    "paddock_name": paddock.paddock_name,
                    "nodes_count": len(nodes),
                    "sensor_averages": {},
                },
                status_code=status.HTTP_200_OK
            )

        # Group readings by sensor type and calculate averages
        sensor_data = {}
        for reading_type, reading_value, node_id, capture_id in recent_readings:
            if reading_type not in sensor_data:
                sensor_data[reading_type] = []
            sensor_data[reading_type].append(reading_value)

        # Calculate averages for each sensor type
        sensor_averages = {}
        sensor_details = {}
        for reading_type, values in sensor_data.items():
            avg_value = sum(values) / len(values)
            sensor_averages[reading_type] = round(avg_value, 2)
            sensor_details[reading_type] = {
                "average": round(avg_value, 2),
                "min": round(min(values), 2),
                "max": round(max(values), 2),
                "count": len(values)  # Number of nodes that reported this sensor type
            }

        return JSONResponse(
            content={
                "success": True,
                "paddock_id": paddock_id,
                "paddock_name": paddock.paddock_name,
                "nodes_count": len(nodes),
                "nodes_with_readings": len(set(reading[2] for reading in recent_readings)),
                "sensor_averages": sensor_averages,
                "sensor_details": sensor_details,
            },
            status_code=status.HTTP_200_OK
        )

    except Exception as e:
        return JSONResponse(
            content={
                "success": False,
                "message": "Failed to retrieve sensor averages",
                "error": str(e)
            },
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
        )

@router.patch("/{paddock_id}", tags=["paddock"], description="Update a paddock's name.")
async def update_paddock(
    paddock_id: int,
    paddock_update: PaddockUpdate,
    db: Session = Depends(get_db),
    current_user: db_models.user = Depends(get_current_user)
):
    """
    Updates the name of a paddock for the authenticated user.
    """
    try:
        # Find the paddock and ensure it belongs to the current user
        paddock = db.query(db_models.paddock).filter(
            db_models.paddock.user_id == current_user.user_id,
            db_models.paddock.paddock_id == paddock_id
        ).first()

        if not paddock:
            return JSONResponse(
                content={
                    "success": False,
                    "message": "Paddock not found"
                },
                status_code=status.HTTP_404_NOT_FOUND
            )

        # Check if the new name already exists for this user (excluding current paddock)
        existing_paddock = db.query(db_models.paddock).filter(
            db_models.paddock.user_id == current_user.user_id,
            db_models.paddock.paddock_name == paddock_update.paddock_name,
            db_models.paddock.paddock_id != paddock_id
        ).first()

        if existing_paddock:
            return JSONResponse(
                content={
                    "success": False,
                    "message": f"A paddock named '{paddock_update.paddock_name}' already exists"
                },
                status_code=status.HTTP_409_CONFLICT
            )

        # Update the paddock name
        paddock.paddock_name = paddock_update.paddock_name
        db.commit()
        db.refresh(paddock)

        return JSONResponse(
            content={
                "success": True,
                "message": "Paddock updated successfully",
                "paddock": {
                    "paddock_id": paddock.paddock_id,
                    "paddock_name": paddock.paddock_name
                }
            },
            status_code=status.HTTP_200_OK
        )

    except Exception as e:
        db.rollback()
        return JSONResponse(
            content={
                "success": False,
                "message": "Failed to update paddock",
                "error": str(e)
            },
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
        )

@router.delete("/{paddock_id}", tags=["paddock"], description="Delete a paddock and unlink all devices.")
async def delete_paddock(
    paddock_id: int,
    db: Session = Depends(get_db),
    current_user: db_models.user = Depends(get_current_user)
):
    """
    Deletes a paddock and unlinks all associated devices by setting their
    paddock_id and user_id to NULL.
    """
    try:
        # Find the paddock and ensure it belongs to the current user
        paddock = db.query(db_models.paddock).filter(
            db_models.paddock.user_id == current_user.user_id,
            db_models.paddock.paddock_id == paddock_id
        ).first()

        if not paddock:
            return JSONResponse(
                content={
                    "success": False,
                    "message": "Paddock not found"
                },
                status_code=status.HTTP_404_NOT_FOUND
            )

        # Unlink all devices from this paddock
        db.query(db_models.active_node).filter(
            db_models.active_node.paddock_id == paddock_id
        ).update({
            "paddock_id": None,
            "user_id": None
        }, synchronize_session=False)

        # Delete the paddock
        db.delete(paddock)
        db.commit()

        return JSONResponse(
            content={
                "success": True,
                "message": "Paddock deleted successfully and all devices unlinked"
            },
            status_code=status.HTTP_200_OK
        )

    except Exception as e:
        db.rollback()
        return JSONResponse(
            content={
                "success": False,
                "message": "Failed to delete paddock",
                "error": str(e)
            },
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
        )